### DISCUSSION TITLE
How does one specify the architecture when using "load file program.hex" in decomp_dbg

### DISCUSSION THREAD

**Discussion by 46cv8** [NationalSecurityAgency/ghidra#5753] [Q&A]
How does one specify the architecture when using "load file program.hex" in decomp_dbg
I am trying to debug an issue I am having with the decompiler that is similar to the one mentioned in the following ticket.
https://github.com/NationalSecurityAgency/ghidra/issues/4148

As part of doing this I am trying to use the decomp_dbg executable.

I first attempted to export the "Debug function decompilation" file from both a working function and a broken function.

The result was the following error for both of them.
```
[decomp]> restore FUN_0083a4_initialize_HELP
Execution error: Missing name and id attributes in scope
[decomp]> help
```

I assumed this was because the decomp_dbg was missing the entire executable context. (I could be wrong about this as I couldn't find anything explaining what might cause this error)
So I then proceeded to attempt to decompile the entire image file for this stm8 however I am unable to determine how to specify the architecture to the "load file" directive and get the following error.

```
[decomp]> load file program_memory.hex
Cannot convert bfd target to sleigh target: unknown:ihex
Could not create architecture
```

Looking at the source code for loadimage_bfd.cc it looks like there should be some way to supply a "target" but I'm not sure how this can be specified when using the command line interface for "decomp_dbg". 
https://github.com/NationalSecurityAgency/ghidra/blob/d7d6b44e296ac4a215766916d5c24e8b53b2909a/Ghidra/Features/Decompiler/src/decompile/cpp/loadimage_bfd.cc#L23C25-L23C25

If it can't be provided, is there some way to export a file from Ghidra which the decomp_dbg executable can load? attempted to "Export  Program" as both an XML file and Ghidra Zip File. But neither were recognized by "decomp_dbg" "load file".

**Comment by 46cv8** [comment]
After looking at https://github.com/NationalSecurityAgency/ghidra/blob/d7d6b44e296ac4a215766916d5c24e8b53b2909a/Ghidra/Features/Decompiler/src/decompile/cpp/consolemain.cc#L211 it seems that the first argument is the target and the second the file. However this seems to be giving me a strange error.

```
[decomp]> load file STM8:BE:24:default:default program_memory.hex
Unable to open image file: program_memory.hex
Could not create architecture
```

It obviously is finding the image file as it found it just fine when I didn't specify the target. But it seems it must be having some trouble with the architecture. I did include the architecture in the "ghidra_10.3.3_PUBLIC/Ghidra/Processors" folder for STM8 loaded from https://github.com/esaulenka/ghidra_STM8 (and this does work and decompile some functions successfully from within ghidra). Also I did set the "SLEIGHHOME=~/ghidra_10.3.3_PUBLIC" so I know that isn't an issue either.

I guess I'll spend more time looking through the source code to see if perhaps my format for the "target" is incorrect. Yeah it seems that bfd_openr isn't able to find the target even when I specify it
[  thebfd = bfd_openr(filename.c_str(),target.c_str());](https://github.com/NationalSecurityAgency/ghidra/blob/d7d6b44e296ac4a215766916d5c24e8b53b2909a/Ghidra/Features/Decompiler/src/decompile/cpp/loadimage_bfd.cc#L78)

Looking in the source code for bfd_openr it isn't really clear to me how it's building it's bfd_target_vector, but it doesn't seem to be picking up the targets that Ghidra normally accepts.
https://github.com/redox-os/binutils-gdb/blob/master/bfd/targets.c#L1475

**Comment by astrelsky** [comment]
In order to use decomp_debug you must export the decompiled function from the gui. In the gui there is a down arrow in the decompiler and an option to export it to xml. Then in decomp_debug you use `restore file xml` `lo fu functionname`

**Reply by 46cv8** [reply]
Thanks for your reply. I attempted this approach initially as I mentioned in my question.

The result was the following.
```
[decomp]> restore FUN_0083a4_initialize_HELP
Execution error: Missing name and id attributes in scope
```
(this fuinction failed to decompile in Ghidra)

Just incase the issue was related to the fact the function was not decompilable due to some other issue.
I chose a function that decompiled correctly under Ghidra, but I got the same error.

```
[decomp]> restore FUN_00810b_irq_reset.xml
Execution error: Missing name and id attributes in scope
```
(this fuinction successfully decompiled in Ghidra)

So for whatever reason there is some issue preventing the decomp_dbg program from getting the same environment as what Ghidra has or something.

I realize this is not the original question from the title though so if you want to close it out I'm totally fine with that.
It just seems strange to me that what I can decompile under Ghidra, when exported will not decompile under decomp_dbg.

As stated in my original question, my assumption was that the exported XML was missing some required context.
This was why I attempted to decompile the whole binary image using load, which lead to me experiencing the issue mentioned in the subject.

**Reply by 46cv8** [reply]
I uploaded both the exported debug files to a file sharing site you can try yourself to see what the issue might be.

https://ufile.io/f/8u0qv

FUN_00810b_irq_reset.xml (decompiles fine in Ghidra)
FUN_0083a4_initialize_HELP.xml (is the one that doesn't decompile even in Ghidra)

**Reply by K1TCH** [reply]
I'm seeing the same execution error output when trying to restore an XML dump

**Reply by 46cv8** [reply]
Thanks K1TCH, given that at least one of these functions (FUN_00810b_irq_reset.xml) can be viewed fine in Ghidra but not under decomp_dbg directly. How would one go about ascertaining the difference in how it is being invoked? I can only assume it is missing element of the environment/context related to either the total binary OR the STM8 plugin. I can post the entire dumped binary too to attempt to run the decompiler against that but I'm pretty sure then you will run into the other issue I mentioned  above with bfd_opener not being able to find the proper architecture as this is not a standard architecture supported by it and it doesn't seem to be able to get the figure it out.

**Reply by 46cv8** [reply]
If you guys @K1TCH or @astrelsky think it would be appropriate to create an bug issue concerning the fact I can't open the exported file in decomp_dbg for the function the decompiles fine in Ghidra UI let me know and I'll summarize that and provide example files. I was hoping to try and figure out what was wrong with the assembly in the function I couldn't decompile in Ghidra UI. So I could tweak the bytecode to get it to decompile (add afew nop's or something) but since I can't even get the decomp_dbg to work on a function that decompiles fine in the Ghidra UI I'm at a bit of a road block.
